### PR TITLE
Unit Tests

### PR DESCRIPTION
**Pull request by jlberbesi**
Hello fellow code reviewer,

Here the actions we just worked and submitted:

- Added unit tests for all the methods in all the classes: Book, Person, Student, Teacher, Classroom, Rental

### CODE REVIEW

File: student_spec.rb
Comment (shahadat3669) [inline student_spec.rb:1]: - [ ] As per the project requirement, you should place all test files in a separate directory. Currently, some of your test files are in the `spec` folder, but I recommend moving all test files into it. This is a good practice for organizing your project.

- [ ] In this file, you have written test cases for all the methods of the `student.rb` file, except for the `to_hash` method. As per the project requirement, you are expected to add unit tests for all the methods in all the classes. Please consider adding test cases for the `to_hash` method as well to ensure comprehensive test coverage.
- [ ] Additionally in all files (`Book`, `classroom`, `person`, `rental`, `student`, `teacher`) you did not add any test case for the ' to_hash` methods. so I will recommend you go through each of them and add a test case for those methods too.

Completing these steps will help ensure that your project adheres to the organizational structure required and includes tests for all methods, contributing to a more robust and well-organized project. Keep up the good work! 👍🧪📂

File: spec/person_spec.rb
Comment (shahadat3669) [inline spec/person_spec.rb:1]: - [ ] Good work! In this file, you have written test cases for all the methods of the `person.rb` file, except for the `can_use_services?` and  `to_hash` methods. As per the project requirement, you are expected to add unit tests for all the methods in all the classes. Please consider adding test cases for the `can_use_services?` and `to_hash` methods as well to ensure comprehensive test coverage.

Completing these steps will help ensure that your project adheres to the organizational structure required and includes tests for all methods, contributing to a more robust and well-organized project. Keep up the good work! 👍🧪📂

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi  @jlberbesi, @CarlosZ96, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Descriptive PR Title
✅ Descriptive PR Description
✅ The code is top-notch! 💯


## Required Changes ♻️

- [ ] Currently, I am unable to find all the work that you did during the "[OOP school library: decorate a class](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_decorator_v2.md#oop-school-library-decorate-a-class)" milestone. I recommend checking your code and adding back the functionality from the previous milestone. Each milestone builds on the previous one, so code consistency is important. I noticed that during [this](https://github.com/CarlosZ96/-Microverse-School-library/pull/5) milestone, you deleted those methods. Please consider updating your code to maintain consistency and ensure that all required functionality is in place.

- [ ] Additionally, I noticed that you pushed two comments after the review request. To streamline the workflow, it's generally recommended to make changes before requesting a review, if possible. You added some files that may be used during decorator milestone. but they are not tested yet. please configure them properly and add testcase for them.

Ensuring that the functionality is preserved and consistent throughout your project will contribute to a more complete and polished final product. Keep up the good work! 👍🧰🔄

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by shahadat3669** [comment]
@jlberbesi, @CarlosZ96, 👀 While reviewing your project, I noticed that you pushed two new commits with some changes. 📝 It's important to remember that pushing new changes after requesting a review isn't considered a best practice. 🚫 This can make the review process more challenging for the reviewer, as they may need to reevaluate the code with the new changes.

💡 To streamline the review process, I suggest updating your code as comprehensively as possible before making a review request. Once the request is submitted, it's generally best to wait for feedback before making additional changes. 🕒 This approach ensures that the reviewer can provide meaningful feedback based on the state of the code at the time of the request. 🤝

File: student_spec.rb
Comment (CarlosZ96) [reply student_spec.rb:1]: image

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  end
+  describe '#to_hash' do
+    it 'returns a hash with person details' do
+      hash = @person.to_hash
+      expect(hash).to be_a(Hash)
+      expect(hash[:id]).to eq(@person.id)
+      expect(hash[:name]).to eq('John Doe')
+      expect(hash[:age]).to eq(24)
+      expect(hash[:parent_permission]).to eq(true)
+    end
```
Comment (iambenkis) [inline spec/person_spec.rb:52]: - Good work applying the changes requested by the previous reviewer! However, as he requested you should also add tests cases for the can_use_services Methods. As per the project requirement, you are expected to add unit tests for all the methods in all the classes. Please consider adding test cases for the can_use_services to meet the requierements

File: spec/rental_spec.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      expect(rental.person).to eq person
+    end
+  end
+
+  describe 'associations' do
+    it 'associates a book with a person' do
+      expect(book.rentals).to include(rental)
+      expect(person.rentals).to include(rental)
+    end
+  end
```
Comment (iambenkis) [inline spec/rental_spec.rb:23]: - [ ] Rental spec should have test cases  on the   `to_hash`  method 👍 as request by the previous code reviewer , Kindly add it

**Review by iambenkis — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @jlberbesi  
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

almost-there

## To highlight :champagne: 

✅ Descriptive PR Description
✅ The code is top-notch! 💯

## Required changes :recycle: :recycle: 

- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me @iambenkis  when you comment, so I can receive the notification or use Slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/rental_spec.rb (hunk 2 above)
Comment (CarlosZ96) [reply spec/rental_spec.rb:23]: image done!

File: spec/person_spec.rb (hunk 1 above)
Comment (CarlosZ96) [reply spec/person_spec.rb:52]: image done!

**Review by abdulhamiid — APPROVED** [review]
# :green_circle:  Approved! :partying_face:
Hi Team 👋, 
image
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 🌟

- ✔️ Descriptions are descriptive
- ✔️ Write unit tests using Rspec.

## Optional suggestions 🤷‍♂️
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

## Important Notes 📝
- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**